### PR TITLE
[AIRFLOW-1932] Add GCP Pub/Sub Pull and Ack

### PR DESCRIPTION
**Pull request by prodonjs**
### JIRA
- [X] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW-1932) issues and references them in the PR title.

### Description
- [X] Here are some details about my PR, including screenshots of any UI changes:

Adds the necessary hooks to support pulling and acknowleding Pub/Sub messages. This is implemented by adding a PubSubPullSensor operator that will attempt to retrieve messages from a specified subscription and will meet its criteria when a message or messages is available. The configuration allows those messages to be acknowledged immediately. In addition, the messages are passed to downstream workers via the return value of operator's execute method. 

An end-to-end example is included showing topic and subscription creation, parallel tasks to publish and pull messages, and a downstream chain to echo the contents of each message before cleaning up.

### Tests
- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
 - `tests.contrib.sensors.test_pubsub_sensor.py`

### Commits
- [X] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

- [X] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

**Comment by bolkedebruin** [comment]
@TrevorEdwards would you mind taking a look?

**Comment by TrevorEdwards** [comment]
Sure, I'll try to review this by end of today. Note that this has a dependency on #2872

**Comment by TrevorEdwards** [comment]
@fenglu-g Mind also taking a look?

File: airflow/contrib/example_dags/example_pubsub_flow.py
Comment (TrevorEdwards) [inline airflow/contrib/example_dags/example_pubsub_flow.py]: nit: datetime.timedelta(days=1)?
Or have a comment here for the user to fill in their own desired schedule_interval

File: airflow/contrib/example_dags/example_pubsub_flow.py
Comment (TrevorEdwards) [inline airflow/contrib/example_dags/example_pubsub_flow.py]: nit: using the 

`with (...) as dag:`

syntax will save you from needing to repeat dag=dag on every operator

**Review by TrevorEdwards — APPROVED** [review]
Note I only reviewed the changes that are separate from #2872.

File: airflow/contrib/example_dags/example_pubsub_flow.py
Comment (prodonjs) [reply airflow/contrib/example_dags/example_pubsub_flow.py]: Excellent. I also assume I could have added project_id and topic to to default_args to avoid repeated those args as well correct? As with the prior comment, would you prefer that I add those prior to your merge?

File: airflow/contrib/example_dags/example_pubsub_flow.py
Comment (prodonjs) [reply airflow/contrib/example_dags/example_pubsub_flow.py]: Understood. Would you prefer I change this prior to when you'll merge?

Comment (TrevorEdwards) [reply airflow/contrib/example_dags/example_pubsub_flow.py]: Yes, please make both changes prior to merge. I usually make the changes and git rebase so that there is still one clean commit. I am just a contributor, not a committer so I will not be making the merge, one of the committers (probably Bolke) will take a look and decide whether to merge :)

File: airflow/contrib/example_dags/example_pubsub_flow.py
Comment (TrevorEdwards) [reply airflow/contrib/example_dags/example_pubsub_flow.py]: Believe you are right about project_id and topic, feel free to make that change too. As said before, please change before merge.

File: airflow/contrib/hooks/gcp_pubsub_hook.py
Comment (fenglu-db) [inline airflow/contrib/hooks/gcp_pubsub_hook.py]: nit: format method is preferred for consistency and flexibility, here and other places.

File: airflow/contrib/sensors/pubsub_sensor.py
Comment (fenglu-g) [inline airflow/contrib/sensors/pubsub_sensor.py]: please don't override the execute implementation unless your polling logic differs from the default:  https://github.com/apache/incubator-airflow/blob/master/airflow/operators/sensors.py#L77

File: airflow/contrib/operators/pubsub_operator.py
Comment (fenglu-db) [inline airflow/contrib/operators/pubsub_operator.py]: return hook.create_subscription(...) to make it auto push to XCOM, otherwise an anonymous subscription is not useful, fine with forcing non-empty subscription on the other hand.

**Review by fenglu-db — COMMENTED** [review]
Mostly LGTM, please also add enough logging in pubsub_hook to facilitate debugging.

**Comment by bolkedebruin** [comment]
Please rebase and fix last issues, then I can merge.

**Comment by prodonjs** [comment]
PTAL. All issues should be addressed. With respect to logging, the base Google API client provides decent details on the HTTP requests being made and the hook methods should log any unexpected failures.

**Comment by fenglu-db** [comment]
I could be wrong, but I don't think the Google API client's log/print message will appear in Airflow log. Otherwise LGTM.

**Comment by prodonjs** [comment]
If entries appear in task logs, would that also appear in the Airflow log? For instance, the create-subscription task log from my run of the pubsub-end-to-end dag shows the following..

```
[2017-12-19 17:10:49,036] {logging_mixin.py:90} INFO - [2017-12-19 17:10:49,036] {gcp_api_base_hook.py:72} INFO - Getting connection using `gcloud auth` user, since no key file is defined for hook.
[2017-12-19 17:10:49,047] {logging_mixin.py:90} INFO - [2017-12-19 17:10:49,047] {discovery.py:852} INFO - URL being requested: PUT https://pubsub.googleapis.com/v1/projects/<project-id>/subscriptions/subscription-to-example-topic?alt=json
[2017-12-19 17:10:49,048] {logging_mixin.py:90} INFO - [2017-12-19 17:10:49,047] {client.py:595} INFO - Attempting refresh to obtain initial access_token
[2017-12-19 17:10:49,048] {logging_mixin.py:90} INFO - [2017-12-19 17:10:49,048] {client.py:893} INFO - Refreshing access_token
[2017-12-19 17:11:21,397] {logging_mixin.py:90} INFO - [2017-12-19 17:11:21,397] {jobs.py:2510} INFO - Task exited with return code 0
```

**Comment by fenglu-db** [comment]
Sorry, I missed that part. LGTM, thanks Jason!

**Comment by bolkedebruin** [comment]
Thanks for the PR! Merged.

**Pull request closed without merging**
